src/Utilities/functions.jl

```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
-    for t in sorted_terms
-        if !isempty(terms) && t.variable_index == last(terms).variable_index
-            terms[end] = MOI.ScalarAffineTerm(terms[end].coefficient + t.coefficient, t.variable_index)
-        elseif !iszero(t.coefficient)
-            if !isempty(terms) && iszero(last(terms).coefficient)
-                terms[end] = t
-            else
-                push!(terms, t)
-            end
-        end
+function canonical(f::MOI.ScalarAffineFunction)
+    if iscanonical(f)
+        f
+    else
+        canonical!(MOI.ScalarAffineFunction(copy(f.terms), f.constant))
     end
-    if !isempty(terms) && iszero(last(terms).coefficient)
-        pop!(terms)
+end
+
+function canonical(f::MOI.VectorAffineFunction)
+    if iscanonical(f)
+        f
+    else
+        canonical!(MOI.VectorAffineFunction(copy(f.terms), f.constants))
     end
-    SAF{T}(terms, f.constant)
 end
-function canonical(f::VAF{T}) where T
-    sorted_terms = sort(f.terms, by = t -> (t.output_index, t.scalar_term.variable_index.value))
-    terms = MOI.VectorAffineTerm{T}[]
-    for t in sorted_terms
-        if !isempty(terms) && t.output_index == last(terms).output_index && t.scalar_term.variable_index == last(terms).scalar_term.variable_index
-            terms[end] = MOI.VectorAffineTerm(t.output_index, MOI.ScalarAffineTerm(terms[end].scalar_term.coefficient + t.scalar_term.coefficient, t.scalar_term.variable_index))
-        elseif !iszero(t.scalar_term.coefficient)
-            if !isempty(terms) && iszero(last(terms).scalar_term.coefficient)
-                terms[end] = t
+
+
+"""
+    canonical!(f::AbstractFunction)
+
+Convert a function to canonical form in-place, without allocating a copy to hold the result.
+See [`canonical`](@ref).
+"""
+function canonical!(f::MOI.ScalarAffineFunction)
+    merge_duplicates!(f.terms,
+                      t -> t.variable_index.value,
+                      t -> !iszero(t.coefficient),
+                      (t1, t2) -> MOI.ScalarAffineTerm(t1.coefficient + t2.coefficient, t1.variable_index))
+    f
+end
+
+
+function canonical!(f::MOI.VectorAffineFunction)
+    merge_duplicates!(f.terms,
+                      t -> (t.output_index, t.scalar_term.variable_index.value),
+                      t -> !iszero(t.scalar_term.coefficient),
+                      (t1, t2) -> MOI.VectorAffineTerm(t1.output_index,
+                                                       MOI.ScalarAffineTerm(t1.scalar_term.coefficient + t2.scalar_term.coefficient,
+                                                                            t1.scalar_term.variable_index)))
+    f
+end
+
+function merge_duplicates!(x::AbstractVector, by, keep, combine)
+    if length(x) == 1
+        if !keep(first(x))
+            empty!(x)
+        end
+    elseif length(x) > 1
+        sort!(x, QuickSort, Base.Order.ord(isless, by, false, Base.Sort.Forward))
+        i1 = 1
+        i2 = 2
+        @inbounds while i2 <= length(x)
```

Comment:
`i2` gets incremented in both branches, so I'd just use `for i2 in eachindex(x)[2 : end]`.

Reply:
fixed